File: Python/CylindricalCapacitorFEM.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.tri as mtri
import scipy.sparse as sp

a = 0.76
b = 1.75
V = 1


# Create Geometry
R = np.linspace(a, b, 50)
Theta = np.linspace(0, 2*np.pi, 100)

R_matrix, Theta_matrix = np.meshgrid(R, Theta)
X_n = R_matrix * np.cos(Theta_matrix)
Y_n = R_matrix * np.sin(Theta_matrix)
X = X_n.reshape(X_n.size)
Y = Y_n.reshape(Y_n.size)

triang = mtri.Triangulation(X, Y)
tri_coords = np.zeros((triang.triangles.shape[0], 3, 2))
for i, tri in enumerate(triang.triangles):
    tri_coords[i] = np.column_stack((X[tri], Y[tri]))

mask = np.zeros(triang.triangles.shape[0], dtype=bool)
for i in range(0, triang.triangles.shape[0]):
    t_c = tri_coords[i]
    x1 = t_c[0][0]
    x2 = t_c[1][0]
    x3 = t_c[2][0]
    y1 = t_c[0][1]
    y2 = t_c[1][1]
    y3 = t_c[2][1]
    r1 = np.sqrt(pow(x1, 2) + pow(y1, 2))
    r2 = np.sqrt(pow(x2, 2) + pow(y2, 2))
    r3 = np.sqrt(pow(x3, 2) + pow(y3, 2))
    if r1-a <= pow(10,-6) and r2-a <= pow(10,-6) and r3-a <= pow(10,-6):
        mask[i] = True
triang.set_mask(mask)

nodes = np.column_stack((triang.x, triang.y))
unique_nodes = np.unique(nodes, axis=0)
element_nodes = triang.triangles


# Define known and unknown potentials
node_id = np.ones(nodes.shape[0])
Potentials = np.zeros(nodes.shape[0])
for inode in range(0, nodes.shape[0]):
    radius = np.sqrt(pow(nodes[inode][0],2) + pow(nodes[inode][1], 2))
    if abs(radius-a) <= pow(10,-6):
        node_id[inode] = 0
        Potentials[inode] = V
    elif abs(radius-b) <= pow(10,-6):
        node_id[inode] = 0
        Potentials[inode] = 0

# Index of unkowns
un_index = np.zeros(nodes.size, dtype=int)
counter = 0
for inode in range(0, nodes.shape[0]):
    if node_id[inode] == 1:
        un_index[inode] = counter
        counter = counter + 1
      


# Matrix Calculation
Nf = counter
Sff = sp.lil_matrix((Nf, Nf), dtype = float)
B = np.zeros(Nf)


for ie in range(1, triang.triangles.shape[0]):
    t_c = tri_coords[ie]
    x = np.zeros(3)
    y = np.zeros(3)
    x = [t_c[0][0], t_c[1][0], t_c[2][0]]
    y = [t_c[0][1], t_c[1][1], t_c[2][1]]
    n = [element_nodes[ie][0], element_nodes[ie][1], element_nodes[ie][2]]
        
    A = np.array([[1, x[0], y[0]], [1, x[1], y[1]], [1, x[2], y[2]]])
    De = np.linalg.det(A)
    Ae = abs(De/2)
    b = np.zeros(3)
    c = np.zeros(3)
    b = [(y[1]-y[2])/De, (y[2]-y[0])/De, (y[0]-y[1])/De]
    c = [(x[2]-x[1])/De, (x[0]-x[2])/De, (x[1]-x[0])/De]

    Se = np.zeros([3, 3])
    for i in range(0, 3):
        for j in range(0, 3):
            Se[i][j] = (b[i]*b[j] + c[i]*c[j])*Ae

            if node_id[n[i]] == 1:
                if node_id[n[j]] == 1:
                    Sff[un_index[n[i]], un_index[n[j]]] = Sff[un_index[n[i]], un_index[n[j]]] + Se[i][j]
                else:
                    B[un_index[n[i]]] = B[un_index[n[i]]] - Se[i][j]*Potentials[n[j]]


# Unkown Potential Calculation
P = sp.linalg.bicgstab(Sff, B)
pot = P[0]
for inode in range(0, nodes.shape[0]):
    if node_id[inode] == 1:
        Potentials[inode] = pot[un_index[inode]]


# Plot the results
fig, ax = plt.subplots()
ax.triplot(triang)
cax = ax.tripcolor(triang, Potentials, cmap='plasma', shading='flat')
fig.colorbar(cax, ax=ax)
ax.set_xlabel('x')
ax.set_ylabel('y')
plt.show()

File: MATLAB/CylindricalCapacitorFEM.m
clear
clc

a = 0.76;
b = 1.75;
V = 1;

% Geometry Description
gd = [1 1; 0 0; 0 0; b a];
ns = [82 82; 49 50];
dl = decsg(gd, 'R1-R2', ns);

% Create Mesh
[p, e, t] = initmesh(dl);
[p, e, t] = refinemesh(dl, p, e, t);
[p, e, t] = refinemesh(dl, p, e, t);
[p, e, t] = refinemesh(dl, p, e, t);
Num_nodes = size(p, 2);
Num_elements = size(t, 2);


% Define known and unknown potentials
node_id = ones(Num_nodes, 1);
Potentials = zeros(Num_nodes, 1);
Num_edges = size(e, 2);
for ie = 1:Num_edges
    region1 = e(6, ie);
    region2 = e(7, ie);
    nodes(1:2) = e(1:2, ie);
    x(1:2) = p(1, nodes(1:2));
    y(1:2) = p(2, nodes(1:2));
    radius(1:2) = sqrt(x(1:2).^2 + y(1:2).^2);

    if (region1 == 0 || region2 == 0)
        node_id(nodes(1)) = 0;
        node_id(nodes(2)) = 0;
        
        if (radius(1) < (a+b)/2)
            Potentials(nodes(1)) = V;
        else
            Potentials(nodes(1)) = 0;
        end
        if (radius(2) < (a+b)/2)
            Potentials(nodes(2)) = V;
        else
            Potentials(nodes(2)) = 0;
        end
    end
end

% Indexes of unknowns
non_pot_index = zeros(Num_nodes, 1);
counter = 0;
for in = 1:Num_nodes
    if (node_id(in) == 1)
        counter = counter + 1;
        non_pot_index(in) = counter;
    end
end

% Matrix Calculation
Nf = counter;
Sff = spalloc(Nf, Nf, 7*Nf);
B = zeros(Nf, 1);

for ie = 1:Num_elements
    nodes(1:3) = t(1:3, ie);
    region = t(4, ie);
    x(1:3) = p(1, nodes(1:3));
    y(1:3) = p(2, nodes(1:3));

    De = det([1 x(1) y(1); 1 x(2) y(2); 1 x(3) y(3)]);
    Ae = abs(De/2);
    b(1) = (y(2) - y(3))/De;
    b(2) = (y(3) - y(1))/De;
    b(3) = (y(1) - y(2))/De;
    c(1) = (x(3) - x(2))/De;
    c(2) = (x(1) - x(3))/De;
    c(3) = (x(2) - x(1))/De;

    for i = 1:3
        for j = 1:3
            Se(i, j) = (b(i)*b(j) + c(i)*c(j))*Ae;

            if (node_id(nodes(i)) == 1)
                if (node_id(nodes(j)) == 1)
                    Sff(non_pot_index(nodes(i)), non_pot_index(nodes(j))) = ...
                    Sff(non_pot_index(nodes(i)), non_pot_index(nodes(j))) + Se(i, j);
                else
                    B(non_pot_index(nodes(i))) = B(non_pot_index(nodes(i))) - ...
                    Se(i, j)*Potentials(nodes(j));
                end
            end
        end
    end
end

% Unknown Potential Calculation 
pot = Sff\B;
for in = 1:Num_nodes
    if node_id(in) == 1
        Potentials(in) = pot(non_pot_index(in));
    end
end

% Calculate Electric Field
[Ex, Ey] = pdegrad(p, t, Potentials);
Ex = -Ex;
Ey = -Ey;

% Plot the Results
figure(1);
pdeplot(p, e, t, 'xydata', Potentials);
axis tight;
axis equal;
hold on;
pdeplot(p, e, t, 'FlowData', [Ex', Ey']');
colormap jet;

